boa/contracts/abi/contract.py

```diff
@@ -0,0 +1,137 @@
+        self._functions = functions
+
+        for name, group in groupby(self._functions, key=attrgetter("name")):
+            setattr(self, name, ABIOverload.create(list(group), self))
+
+        self._address = Address(address)
+
+    @cached_property
+    def method_id_map(self):
+        """
+        Returns a mapping from method id to function object.
+        This is used to create the stack trace when an error occurs.
+        """
+        return {function.method_id: function for function in self._functions}
+
+    def marshal_to_python(self, computation, abi_type: list[str]) -> tuple[Any, ...]:
+        """
+        Convert the output of a contract call to a Python object.
+        :param computation: the computation object returned by `execute_code`
+        :param abi_type: the ABI type of the return value.
+        """
+        if computation.is_error:
+            return self.handle_error(computation)
+
+        schema = f"({','.join(abi_type)})"
+        decoded = abi_decode(schema, computation.output)
+        return tuple(_decode_addresses(typ, val) for typ, val in zip(abi_type, decoded))
+
+    def stack_trace(self, computation: ComputationAPI) -> StackTrace:
+        """
+        Create a stack trace for a failed contract call.
+        """
+        calldata_method_id = bytes(computation.msg.data[:4])
+        if calldata_method_id in self.method_id_map:
+            function = self.method_id_map[calldata_method_id]
+            msg = f"  (unknown location in {self}.{function.full_signature})"
+        else:
+            # Method might not be specified in the ABI
+            msg = f"  (unknown method id {self}.0x{calldata_method_id.hex()})"
+
+        return_trace = StackTrace([msg])
+        return _handle_child_trace(computation, self.env, return_trace)
+
+    @property
+    def deployer(self) -> "ABIContractFactory":
+        """
+        Returns a factory that can be used to retrieve another deployed contract.
+        """
+        return ABIContractFactory(self._name, self._functions)
+
+    def __repr__(self):
+        file_str = f" (file {self.filename})" if self.filename else ""
+        return f"<{self._name} interface at {self.address}>{file_str}"
+
+
+class ABIContractFactory:
+    """
+    Represents an ABI contract that has not been coupled with an address yet.
+    This is named `Factory` instead of `Deployer` because it doesn't actually
+    do any contract deployment.
+    """
+
+    def __init__(
+        self, name: str, functions: list["ABIFunction"], filename: Optional[str] = None
+    ):
+        self._name = name
+        self._functions = functions
+        self._filename = filename
+
+    @classmethod
+    def from_abi_dict(cls, abi, name: Optional[str] = None):
+        if name is None:
+            name = "<anonymous contract>"
+
+        functions = [
+            ABIFunction(item, name) for item in abi if item.get("type") == "function"
+        ]
+
+        # warn on functions with same name
```

Comment:
are you sure we need a warning here? since one of the main goals of this PR is to handle overloaded functions

Reply:
Definitely not! I'll delete.